FlxSound: add loopTime, closes #1662

winterismute opened the pull request:
Also, tiny cleanup of some FlxSound things.
Related for example to request https://github.com/HaxeFlixel/flixel/issues/1662

Gama11 commented on flixel/system/FlxSound.hx:
```diff
@@ -144,6 +144,10 @@ class FlxSound extends FlxBasic
+	/**
+	 * In case of looping, from where to restart the sound when it loops back
+	 */
+	private var loopPoint(default, set):Float;
```
This should be `public` (otherwise what's the point of making it a property with a setter?) and declared after `looping`.

Gama11 replied:
I'm also not sure I like the name... How about `loopTime` (there already is a `time`) or `loopPosition`?

Gama11 replied:
`loopTime` probably makes the most sense?

Gama11 commented on flixel/system/FlxSound.hx:
```diff
@@ -442,11 +447,41 @@ class FlxSound extends FlxBasic
+	 * @param	newloopPoint The new position in the sound, in milliseconds
+	 * @return	The updated value in milliseconds
+	 */
+	public function set_loopPoint(newloopPoint:Float):Float
```
Setters should always be private.

Gama11 commented on flixel/system/FlxSound.hx:
```diff
@@ -442,11 +447,41 @@ class FlxSound extends FlxBasic
+	 * Sets the sound point at which to restart playing after a loop
+	 * @param	newloopPoint The new position in the sound, in milliseconds
+	 * @return	The updated value in milliseconds
+	 */
```
There's no point to having documentation for setters, the docs should go to the property.

Gama11 commented on flixel/system/FlxSound.hx:
```diff
@@ -704,22 +738,19 @@ class FlxSound extends FlxBasic
-		
 		if (!loop && looped && playing)
 		{
 			looped = loop;
```
This line is now redundant.

Gama11 commented on flixel/system/frontEnds/SoundFrontEnd.hx:
```diff
@@ -95,6 +96,7 @@ class SoundFrontEnd
 		music.volume = Volume;
 		music.persist = true;
 		music.group = (Group == null) ? defaultMusicGroup : Group;
+		music.set_loopPoint(LoopPoint);
```
Should never directly call a property setter, only access it via the property.

Gama11 commented on flixel/system/FlxSound.hx:
```diff
@@ -442,11 +447,41 @@ class FlxSound extends FlxBasic
+	 * @param	min	The minute part of the new position eg. 2 in case of 02:32
+	 * @param	sec	The seconds part of the new position eg. 32 in case of 02:32
+	 */
+	public inline function setloopPointMinSec(min:Int, sec:Int)
```
I think we can do without this for now.

Gama11 commented on flixel/system/FlxSound.hx:
```diff
@@ -383,7 +388,7 @@ class FlxSound extends FlxBasic
 	 * @param	ForceRestart	Whether to start the sound over or not.  Default value is false, meaning if the sound is already playing or was paused when you call play(), it will continue playing from its current position, NOT start again from the beginning.
 	 */
-	public function play(ForceRestart:Bool = false):FlxSound
+	public function play(ForceRestart:Bool = false, Position:Float = 0.0):FlxSound
```
Adding `Position` here creates the awkward situation that the argument is not settable from the `SoundFrontEnd` function. We really should add any more arguments to those though if possible though, they already have too many.

It would be much better to implement a setter for `time`. That should be done separately though.

Gama11 replied:
Hm.. at least call it `Time` then for consistency with the public property.

Gama11 commented on flixel/system/FlxSound.hx:
```diff
@@ -442,11 +447,41 @@ class FlxSound extends FlxBasic
+			startSound(pos);
+			return loopPoint;
+		}
+		loopPoint = newloopPoint;
```
This could be moved before the `if` and save a few lines - it's executed in either case.

Gama11 replied:
Right... I'm not sure I really understand that if-logic anyway. Why is it necessary?

winterismute replied:
I think: you need to spot the case in which the sound is playing, is looping, and has no loopPoint (starts from the beginning). In that case, it meant it was played (startSound()) with MAX_INT as number of loops, which means that it will never stop. In this case, you have to update the channel to set the num of loops to be 0, so that after each completed play "stopped" is called and you can call then "play" with the time parameter. It' similar to what "set_looped" does.

Gama11 replied:
Hm, so it's a matter of having to manually manage the looping now... I wonder why we don't always do that then, seems less complicated. Either way, you should:

- create a helper method that handles this to avoid the redundancy with set_looped()
- add a comment to the call here explaning why it's necessary since it doesn't seem super-intuitive

gamedevsam commented on flixel/system/FlxSound.hx:
```diff
@@ -382,8 +387,9 @@ class FlxSound extends FlxBasic
+	 * @param	Time			At which point to start plaing the sound, in milliseconds
 	 */
-	public function play(ForceRestart:Bool = false):FlxSound
+	public function play(ForceRestart:Bool = false, Time:Float = 0.0):FlxSound
```
This ```time``` variable should be named ```delay``` to better reflect its behavior.

gamedevsam replied:
This variable should be renamed ```StartTime``` for clarity.

gamedevsam commented on flixel/system/frontEnds/SoundFrontEnd.hx:
```diff
@@ -79,8 +79,9 @@ class SoundFrontEnd
 	 * @param	Volume		How loud the sound should be, from 0 to 1.
 	 * @param	Looped		Whether to loop this music.
 	 * @param	Group		The group to add this sound to.
+	 * @param	LoopPoint	Point at which the sound will restart when looping back (in case of looping)
```
Should be ```LoopTime```

gamedevsam commented on flixel/system/FlxSound.hx:
```diff
@@ -442,10 +448,25 @@ class FlxSound extends FlxBasic
+		// If the sound is looping, already playing and the looptime WAS 0
+		//	we need to restart the sound and NOT call play on the _sound property
+		// 	with max int as number of loops, but with 0 instead, handling the loop with the event
+		if (looped && playing && loopTime == 0.0)
```
This function could be condensed for brevity:
```
var requiresRestart:Bool = looped && playing && loopTime == 0.0;
loopTime = newloopTime;
if(requiresRestart)
{
	restartSound();
}
return loopTime;
```

gamedevsam commented on flixel/system/FlxSound.hx:
```diff
@@ -523,16 +544,16 @@ class FlxSound extends FlxBasic
+		var numLoops:Int = looped && loopTime == 0.0 ? FlxMath.MAX_VALUE_INT : 0;
 		
-		time = Position;
+		time = Time;
```
This looks weird, perhaps the variable could be renamed ```startTime```.

winterismute commented:
I have fixed the possibility of looping using the second argument of Sound.play(), now we always use the event. In native builds has 0 effect on performance, not sure about flash, although it leads to nicer code.
Fixed the naming as pointed out by @gamedevsam too.

Gama11 commented:
Thanks!
